### DISCUSSION TITLE
Gathering to discuss the main question & to publish a more deep introduction (aka propaedeutics) into the project

### DISCUSSION THREAD

**Discussion by Anton-Latukha** [haskell-nix/hnix#944] [Ideas]
### Preamble

Working on the project, decided that did enough learning of it through the code reading/refactoring. I did enough of it to get an understanding of a whole project and approached some additional deeper understanding boundary, there are information & scope of questions that only John can cover.

There are hard programming themes in the project, that have a deep vast impact on the project, in particular: laziness, recursive schemes, abstracting definitional interpreters - these are the main & advanced programming paradigms the project is built on and they form the design, code, compilation & execution of the project

### Questions

Questions that see particularly beneficial to ask:

  * History: How long you did what and volume of effort HNix took. How long it took to study/implement topics/parts & properties of the system.
  * Shortcomings in/of the design he knows of or think of as important to solve upfront.
  * Thoughts on the idea to migrate HNix to `recursion-schemes` library.
    * Thoughts on the current code (considering there were no changes to the recursion schemes part of it).
    * The desired scope of recursion schemes application in the code.
    * Maybe look particularly into `Utils.Fix1T` & `Value` and give thoughts on how to put them more classically-defaulty/migrate them to `recursion-schemes`.
  * Abstracting Definitional Interpreters brief.
    * Consequences of the use of the ADI on HNix.
      * Assumption of the scope it currently works at (considering there were no changes to it):
  * What companies you know that use Nix and do they use Haskell. (It is of personal interest, since looking for a place to work at, but also it is useful to know to the project contributors & to maintainers).
  * From what he sees, any improvement thoughts to translate.

### Time windows:

John Wiegley response:
> I am free the following slots:
> 
> Schedule in UTC:
> 
>   Mon 20:00-23:00 UTC
>   Tue 20:00-23:00 UTC
>   Wed 03:00-05:00 UTC
>   Thu 20:00-23:00 UTC
>   Fri 03:00-05:00 UTC && 17:00-20:00 UTC
>   
> Schedule in PST:
> 
>   Mon 13:00-16:00 PST
>   Tue 13:00-16:00 PST && 20:00-22:00 PST
>   Wed
>   Thu 13:00-16:00 PST && 20:00-22:00 PST
>   Fri 10:00-13:00 PST
> 
>
> If none of these times work for anyone else, I can juggle some things around to find more available for you.

Marked dates that suit me in the voting.

### Picking a time window

Please vote:

Multichoice is possible. Check the dates, check the timezone.

1-click  - "I vote here."
2-clicks - "If would be needed."
3-clicks - "Please, no."

https://doodle.com/poll/b6k8cwd738eikb9c?utm_source=poll&utm_medium=link

This is the main event thread, in terms that, in the end of initialization, were and would be published here.

---

# The resulting video

From ~3 hours of a dialog, 1/3 of the dialog was constructed to be a part to be recorded as relevant to new contributors:

https://youtu.be/jtr0DHywTtM

Currently it is processing.

**Comment by Anton-Latukha** [comment]
Time was established as `Friday, June 11 {10:00-13:00 PST, 17:00-20:00 UTC}`.

https://us04web.zoom.us/j/5679333830?pwd=OTFQd0ZkSElQUCtUbWtLWXlTZm9ldz09
Meeting ID: 567 933 3830
Passcode: F93fX2

**Comment by layus** [comment]
For future reference, I had a lot of thinking about the adi interpreter and why it prevents optimisations like `maybeThunk` and printing thunks nicely (theoretically we should be able to inspect the expression and the context hidden in the thunk)

But it takes the form of handwritten notes :-/

![hnix-eval-context](https://user-images.githubusercontent.com/632767/121725531-ad61f180-cae9-11eb-84b1-6b086a221dc7.png)

**Reply by Anton-Latukha** [reply]
I need to parse it several more times, to get more from it.

About scope/provenance simplification, Obsidian guys promised & were thinking about it for several years & tried to deliver it several times, see the ideas they almost finished in: https://github.com/haskell-nix/hnix/pull/804. But also, you already reviewed that one 8), so probably you know those already.

**Reply by Anton-Latukha** [reply]
Please, can you open separate reports gradually criticizing the particular pieces of the system & stating ideas/implementation - it would be simpler to understand the problems & their solutions & changes.

As currently, as I understand the main point, you want to communicate: to change the scoping/thunking/type mechanism so it gets simpler & inspectable.

**Comment by Anton-Latukha** [comment]
:arrow_up: Head post was updated with published video recording.

**Comment by layus** [comment]
i had already found the full paper about ADI, but these slides are way easier to process. Dropping the link here for future reference: http://david.darais.com/assets/papers/abstracting-definitional-interpreters/adi.slides.pdf
